src/extract-metadata.js

```diff
@@ -100,6 +110,85 @@ function readBlockConfig($block) {
   return config;
 }
 
+/**
+ * Looks for metadata from a spreadsheet.
+ * @param {string} url The request URL
+ * @param {object} action The action
+ * @return {object} The metadata
+ */
+async function getGlobalMetadata(url, action) {
+  let metaRules = [];
+  if (action) {
+    const {
+      request,
+      downloader,
+      secrets,
+      logger: log,
+    } = action;
+    const {
+      owner, repo, ref,
+    } = request.params || {};
+    try {
+      const headers = {};
+      const token = (secrets && secrets.GITHUB_TOKEN) || (request.headers ? request.headers['x-github-token'] : '');
+      if (token) {
+        headers['x-github-token'] = token;
+      }
+      const res = await downloader.fetch({
```

Comment:
as I said...add an extra step at the beginning of the pipeline where you schedule the download, so that the downloader can start downloading in parallel, so that it doesn't block the pipeline here:

```js
async function fetchMetadata(context, action) {
  const { request, downloader } = action;
  const {
    owner, repo, ref,
  } = request.params || {};
  try {
    // schedule fetch task
    downloader.fetch({
        uri: `https://${ref}--${repo}--${owner}.hlx.page/metadata.json`,
        options: {
          timeout: secrets ? secrets.HTTP_TIMEOUT_EXTERNAL : 20000,
          headers,
        },
        headers,
        errorOn404: false,
      id: 'metadata',
    });
  } catch (err) {
    if (!err.status) {
      throw err;
    }
  }
}

module.exports = fetchMetadata;
```

and  then here:

```js
  const metadataTask = downloader.getTaskById('metadata');
  if (!metadataTask) {
    logger.info('no external metadata. no metadata task scheduled.');
    return;
  }
  const res = await metadataTask;
  if (res.status !== 200) {
    logger.info(`unable to fetch metadata.json: ${res.status}`);
    return;
  }
  ```

and then register the extra step in `pre.js`:

```js
const fetchMetadata = require('./fetch-metadata.js');
...
pre.before.content = fetchMetadata;
module.exports.pre = pre
```

Reply:
@tripodsan I did it as explained above, but the rendering tests fail because `metaTask` is always undefined. Could you help me debug it?